### PR TITLE
[Merged by Bors] - feat(NumberTheory/EllipticDivisibilitySequence): expose the auxiliary sequence

### PR DESCRIPTION
**Pull request by Multramate**
Expose the auxiliary sequences associated to normalised elliptic divisibility sequences along with their corresponding API, and fix the comments under implementation notes. This is needed to define $n$-division polynomials of elliptic curves in a later downstream file as univariate polynomials with the factor of the bivariate $2$-division polynomial omitted.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/EllipticDivisibilitySequence.lean [hunk 1]
```diff
@@ -109,49 +114,129 @@ private def normEDS'' (b c d : R) : ℕ → R
     have h2 : m + 2 < n + 5 := (lt_add_one _).trans h3
     have h1 : m + 1 < n + 5 := (lt_add_one _).trans h2
     if hn : Even n then
-      normEDS'' b c d (m + 4) * normEDS'' b c d (m + 2) ^ 3 * (if Even m then b ^ 4 else 1) -
-        normEDS'' b c d (m + 1) * normEDS'' b c d (m + 3) ^ 3 * (if Even m then 1 else b ^ 4)
+      preNormEDS' b c d (m + 4) * preNormEDS' b c d (m + 2) ^ 3 * (if Even m then b else 1) -
+      preNormEDS' b c d (m + 1) * preNormEDS' b c d (m + 3) ^ 3 * (if Even m then 1 else b)
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/EllipticDivisibilitySequence.lean:118]:
```suggestion
        preNormEDS' b c d (m + 1) * preNormEDS' b c d (m + 3) ^ 3 * (if Even m then 1 else b)
```
From the [Style Guide](https://leanprover-community.github.io/contribute/style.html): "This rule, i.e., indent an additional two spaces, applies more generally whenever a term spans multiple lines."

File: Mathlib/NumberTheory/EllipticDivisibilitySequence.lean [hunk 2]
```diff
@@ -109,49 +114,129 @@ private def normEDS'' (b c d : R) : ℕ → R
...
     else
       have h5 : m + 5 < n + 5 := add_lt_add_right
         (Nat.div_lt_self (Nat.odd_iff_not_even.mpr hn).pos <| Nat.lt_succ_self 1) 5
-      normEDS'' b c d (m + 2) ^ 2 * normEDS'' b c d (m + 3) * normEDS'' b c d (m + 5) -
-        normEDS'' b c d (m + 1) * normEDS'' b c d (m + 3) * normEDS'' b c d (m + 4) ^ 2
+      preNormEDS' b c d (m + 2) ^ 2 * preNormEDS' b c d (m + 3) * preNormEDS' b c d (m + 5) -
+      preNormEDS' b c d (m + 1) * preNormEDS' b c d (m + 3) * preNormEDS' b c d (m + 4) ^ 2
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/EllipticDivisibilitySequence.lean:123]:
```suggestion
        preNormEDS' b c d (m + 1) * preNormEDS' b c d (m + 3) * preNormEDS' b c d (m + 4) ^ 2
```

File: Mathlib/NumberTheory/EllipticDivisibilitySequence.lean [hunk 3]
```diff
@@ -109,49 +114,129 @@ private def normEDS'' (b c d : R) : ℕ → R
...
+lemma preNormEDS'_three : preNormEDS' b c d 3 = c :=
+  rfl
+
+@[simp]
+lemma preNormEDS'_four : preNormEDS' b c d 4 = d :=
+  rfl
+
+lemma preNormEDS'_odd (m : ℕ) : preNormEDS' b c d (2 * (m + 2) + 1) =
+    preNormEDS' b c d (m + 4) * preNormEDS' b c d (m + 2) ^ 3 * (if Even m then b else 1) -
+    preNormEDS' b c d (m + 1) * preNormEDS' b c d (m + 3) ^ 3 * (if Even m then 1 else b) := by
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/EllipticDivisibilitySequence.lean:149]:
```suggestion
      preNormEDS' b c d (m + 1) * preNormEDS' b c d (m + 3) ^ 3 * (if Even m then 1 else b) := by
```

File: Mathlib/NumberTheory/EllipticDivisibilitySequence.lean [hunk 4]
```diff
@@ -109,49 +114,129 @@ private def normEDS'' (b c d : R) : ℕ → R
...
+  rw [show 2 * (m + 2) + 1 = 2 * m + 5 by rfl, preNormEDS', dif_pos <| even_two_mul _]
+  simp only [m.mul_div_cancel_left two_pos]
+
+lemma preNormEDS'_even (m : ℕ) : preNormEDS' b c d (2 * (m + 3)) =
+    preNormEDS' b c d (m + 2) ^ 2 * preNormEDS' b c d (m + 3) * preNormEDS' b c d (m + 5) -
+    preNormEDS' b c d (m + 1) * preNormEDS' b c d (m + 3) * preNormEDS' b c d (m + 4) ^ 2 := by
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/EllipticDivisibilitySequence.lean:155]:
```suggestion
      preNormEDS' b c d (m + 1) * preNormEDS' b c d (m + 3) * preNormEDS' b c d (m + 4) ^ 2 := by
```

File: Mathlib/NumberTheory/EllipticDivisibilitySequence.lean [hunk 5]
```diff
@@ -109,49 +114,129 @@ private def normEDS'' (b c d : R) : ℕ → R
...
+
 /-- The canonical example of a normalised EDS `W : ℕ → R`,
-with initial values `W(0) = 0`, `W(1) = 1`, `W(2) = b`, `W(3) = c`, and `W(4) = b * d`.
+with initial values `W(0) = 0`, `W(1) = 1`, `W(2) = b`, `W(3) = c`, and `W(4) = d * b`.
 
-This is defined in terms of a truncated sequence whose even terms differ by a factor of `b`. -/
+This is defined in terms of `preNormEDS'` whose even terms differ by a factor of `b`. -/
 def normEDS' (n : ℕ) : R :=
-  normEDS'' b c d n * if Even n then b else 1
+  preNormEDS' (b ^ 4) c d n * if Even n then b else 1
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/EllipticDivisibilitySequence.lean:214]: Why does `b` get replaced by `b^4` here? The definition of `PreNormEDS'` seems to be the same as that of `normEDS''`.

Comment (alreadydone) [reply Mathlib/NumberTheory/EllipticDivisibilitySequence.lean:214]: `b^4` are changed to `b` [earlier in the file](https://github.com/leanprover-community/mathlib4/pull/10814/files#diff-02d4668be95cd41b094c2ec5859acf1cf92c74833b9181a533bcce3a23c00961L112-R118).

**Comment by alreadydone** [comment]
@Multramate Can you merge master and fix the indentation as pointed out by @MichaelStollBayreuth? I'd like to use `preNormEDS` to define $\psi^c_n=\psi_{2n}/\psi_n=(\psi_{n-1}^2\psi_{n+2}-\psi_{n-2}\psi_{n+1}^2)/\psi_2$ in my notes in a division-free way. I think the correct definition is `(preNormEDS (n-1) * preNormEDS (n+2) ^ 2 - preNormEDS (n-2) * preNormEDS (n+1) ^ 2) * (if Even n then 1 else b)` where `preNormEDS` is applied to the parameters `b^4, c, d`.

File: Mathlib/NumberTheory/EllipticDivisibilitySequence.lean [hunk 6]
```diff
@@ -189,43 +272,50 @@ lemma normEDS'_even (m : ℕ) : normEDS' b c d (2 * (m + 3)) * b =
     ring1
 
 /-- The canonical example of a normalised EDS `W : ℤ → R`,
-with initial values `W(0) = 0`, `W(1) = 1`, `W(2) = b`, `W(3) = c`, and `W(4) = b * d`.
+with initial values `W(0) = 0`, `W(1) = 1`, `W(2) = b`, `W(3) = c`, and `W(4) = d * b`.
 
 This extends `normEDS'` by defining its values at negative integers. -/
-def normEDS (n : ℤ) : R := n.sign * normEDS' b c d n.natAbs
+def normEDS (n : ℤ) : R :=
+  n.sign * normEDS' b c d n.natAbs
```
Comment (alreadydone) [inline Mathlib/NumberTheory/EllipticDivisibilitySequence.lean:279]: Can't you define
```
def normEDS (n : ℕ) : R :=
  preNormEDS (b ^ 4) c d n * if Even n then b else 1
```
and remove `normEDS'` altogether? You can always recover `normEDS' n` as `normEDS (ofNat n)`.

**Review by alreadydone — APPROVED** [review]

**Review by alreadydone — APPROVED** [review]
Thanks!
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**